Solved: 백준_시간초과

2ssue opened the pull request:
# 백준 11332. 시간초과 #14 

[문제 링크](https://www.acmicpc.net/problem/11332)

| 난이도 | 정답률(\_%) |
| :----: | :---------: |
| Silver 4 | 25.995% |

## 설계
- 총 테스트 케이스 C
- 내부 입력 최대 범위 N
- 내부 테스트 케이스 횟수 T
- 내부 케이스 제한시간 L
### 시간 복잡도
C, 최대 100번. O(1)
### 공간 복잡도
- BigInteger를 사용했는데 내부적으로 int형 배열을 사용해서 큰 수를 저장할 수 있다.
- 정확히 배열을 어떻게 써서 저장하는지는 잘 모르겠지만 한 칸에 2^32 크기만큼 저장되는 것으로 보임. 따라서 공간 복잡도는 연산의 결과 값 / 2^32 개 😕
![image](https://user-images.githubusercontent.com/42017052/82202536-038f5000-993d-11ea-94be-df615dc5641a.png)
## 풀이
- 연산 횟수를 계산해서 L * 100000000번을 넘으면 TLE로 처리했다.
- 다만 팩토리얼의 경우 계산 횟수가 너무 많아질 수 있으므로, 중간에 L * 100000000 을 넘을 경우 계산은 중단하고 결과를 출력하도록 했다.

2ssue committed:
Solved: 백준_시간초과

changicho commented on problems/baekjoon/11332/sujeong.java:
```diff
@@ -0,0 +1,67 @@
+            BigInteger executeCount = new BigInteger(situation[2]);
+            BigInteger timeLimit = BigInteger.valueOf(Integer.parseInt(situation[3]) * 100000000);
+
+            if (bigO.equals("O(N)")) {
```
Java에서 switch문으로 string을 비교하는 건 어떨까요?

C는 string을 switch문에서 지원하지 않아서 if else를 사용하는 편이 제일 좋지만, java 7 이후부터는 string을 switch에서 분기처리 할 수 있어서요.

2ssue replied:
오 세상에 그렇네요! 옛날엔 char형만 됐던걸로 기억해서 사용하지 않았는데 얼마나 고대 시절의 Java를 기억하고 있던걸까요... 알려주셔서 감사합니다 👍 👍 👍

changicho commented on problems/baekjoon/11332/sujeong.java:
```diff
@@ -0,0 +1,67 @@
+                    continue;
+                }
+            } else if (bigO.equals("O(N^3)")) {
+                if (inputLength.pow(3).multiply(executeCount).compareTo(timeLimit) == 1) {
```
BigInteger는 3제곱까지 제대로 계산하는군요. 

c의 long long 형으로는 3제곱시 overflow가 나서 이부분도 분기처리했어야 했습니다.

2ssue replied:
저도 처음에 long을 사용했다가 찾아보니 자바에는 `BigInteger` 클래스가 있더라구요! python은 형에 크기 제한이 없다고 하는데 C++과 파이썬 코드도 얼른 보고싶습니다 😋

changicho commented on problems/baekjoon/11332/sujeong.java:
```diff
@@ -0,0 +1,67 @@
+                int n = inputLength.intValue();
+
+                while (n-- > 1) {
+                    inputLength = inputLength.multiply(new BigInteger(Integer.toString(n)));
```
BigInteger의 연산은 BitInteger 끼리만 가능하군요.

매번 new를 선언하는것이 불리할 수도 있겠네요.

2ssue replied:
넵 그래서 성능이 많이 안좋다고 합니다. 내부도 배열형태로 저장해서 숫자가 많이 커지면 더 안좋아질 것 같아요. 저도 쓰면서 이 부분이 좀 별로라고 생각했습니다..

changicho commented on problems/baekjoon/11332/sujeong.java:
```diff
@@ -0,0 +1,67 @@
+                }
+            }
+
+            bw.write("May Pass.\n");
```
boolean 형식의 isPass를 두고 여기서 분기처리한다면 위 else if 문들의 continue를 없앨 수 있겠습니다.

나중에 분기가 추가되었을 때 continue문이 없어 발생하는 오류를 구조상 막을 수 있겠네요.

2ssue replied:
네 더 좋은 구조로 한번 고쳐봐야겠네요 👍

changicho reviewed (approved):
java는 BigInteger를 이용해 더 쉽게 문제를 풀 수 있군요.

2ssue committed:
Refactor: 계산 부분 함수화 및 비교 로직 변경

- String을 switch로 비교할 수 있다는 것을 알게 되어 방식 변경
- continue 보다는 early return 문을 사용할 수 있도록 함수화
